File: js/src/formats/report.ts
/** What a conversion could not represent, reported rather than discarded.
 *
 * A token carries more spectrum-level structure than any text format it can be
 * written to, so every export except mzML drops something. The contract is that
 * a conversion never silently invents or discards semantics: each writer returns
 * what it wrote together with what it could not, mirroring the Python
 * ConversionResult so both implementations report the same omissions. */

export type Severity = "warning" | "info"

export interface ConversionIssue {
  code: string
  path: string
  message: string
  severity: Severity
}

export class ConversionResult {
  text = ""
  readonly format: string
  readonly issues: ConversionIssue[] = []

  constructor(format: string, text = "") {
    this.format = format
    this.text = text
  }

  add(code: string, path: string, message: string, severity: Severity = "warning"): void {
    this.issues.push({ code, path, message, severity })
  }

  /** Messages for information the output does not carry. */
  get omitted(): string[] {
    return this.issues.filter(i => i.severity === "warning").map(i => i.message)
  }

  /** True when nothing present in the token was left out. */
  get lossless(): boolean {
    return this.omitted.length === 0
  }

  /** A short report, empty when nothing was dropped. */
  summary(): string {
    const dropped = this.omitted
    if (!dropped.length) return ""
    return [`Not represented by ${this.format.toUpperCase()}:`, ...dropped.map(m => `  - ${m}`)].join("\n")
  }
}

File: src/spectrl/formats/_report.py
"""What a conversion could not represent, reported rather than discarded.

A token carries more spectrum-level structure than any of the text formats it
can be written to, so every export except mzML drops something. The format
contract is that a conversion never silently invents or discards semantics, so
each writer returns what it wrote together with what it could not, using the
same issue shape as mzml.conversion_report: code, path, message, severity.
"""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class ConversionResult:
    """Converted text plus an inventory of what the target format cannot hold.

    Attributes:
        text: The converted document.
        format: Short format name, such as "mzml", "mgf", or "ms2".
        issues: One entry per omission or substitution. Severity is "warning"
            when information present in the token is absent from the output,
            and "info" when something was represented differently but not lost.
    """

    text: str
    format: str
    issues: list[dict] = field(default_factory=list)

    def add(self, code: str, path: str, message: str, severity: str = "warning") -> None:
        self.issues.append({"code": code, "path": path, "message": message, "severity": severity})

    @property
    def omitted(self) -> list[str]:
        """Messages for information the output does not carry."""
        return [i["message"] for i in self.issues if i["severity"] == "warning"]

    @property
    def lossless(self) -> bool:
        """True when nothing present in the token was left out."""
        return not self.omitted

    def summary(self) -> str:
        """A short report for a command line, empty when nothing was dropped."""
        if not self.omitted:
            return ""
        lines = [f"Not represented by {self.format.upper()}:"]
        lines += [f"  - {message}" for message in self.omitted]
        return "\n".join(lines)
